lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+    auto port = ports[i];
+    auto argAttr = mlir::function_like_impl::getArgAttrs(op, i);
+    if (port.annotations.empty()) {
+      newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), argAttr));
+      continue;
+    }
+
+    auto remainingAnnotations = handleRefLike(op, port.annotations, port.type);
+
+    // Overwrite the annotations value with remaining annotations.
+    SmallVector<NamedAttribute> newArgAttr;
+    for (auto argIter = argAttr.begin(), argEnd = argAttr.end();
+         argIter != argEnd; ++argIter) {
+      if (argIter->first != "firrtl.annotations") {
+        newArgAttr.push_back(*argIter);
+        continue;
+      }
+      newArgAttr.push_back({argIter->first, remainingAnnotations});
+    }
+    newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), newArgAttr));
+  }
+
+  return ArrayAttr::get(op->getContext(), newArgAttrs);
+};
+
+void GrandCentralPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  AnnotationSet annotations(circuitOp);
+  if (annotations.empty())
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  // Store a mapping of interface name to InterfaceOp.
+  llvm::StringMap<sv::InterfaceOp> interfaces;
+
+  // Track the order that interfaces should be emitted in.
+  llvm::SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+
+  // Examine the Circuit's Annotations doing work to remove Grand Central
+  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
+  // Annotations with these when done.
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+  for (auto anno : annotations) {
+    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType")) {
+      unprocessedAnnos.push_back(anno.getDict());
+      continue;
+    }
+
+    AugmentedBundleType bundle;
+    if (auto maybeBundle = decodeBundleType(anno))
+      bundle = maybeBundle.getValue();
+    else {
+      emitError(circuitOp.getLoc(),
+                "'firrtl.circuit' op contained an 'AugmentedBundleType' "
+                "Annotation which did not conform to the expected format")
+              .attachNote()
+          << "the problematic 'AugmentedBundleType' is: '" << anno.getDict()
+          << "'";
+      return signalPassFailure();
+    }
+
+    for (auto elt : bundle.elements) {
+      std::pair<StringRef, StringRef> key = {bundle.defName, elt.name};
+      interfaceMap[key] = {elt.tpe, elt.description};
+      interfaceKeys.push_back(key);
+    }
+
+    // If the interface already exists, don't create it.
+    if (interfaces.count(bundle.defName))
+      continue;
+
+    // Create the interface.  This will be populated later.
+    interfaces[bundle.defName] =
+        builder.create<sv::InterfaceOp>(circuitOp->getLoc(), bundle.defName);
+  }
+
+  circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
```

Comment:
There's now `annotations.removeAnnotations(...)` that allows you to put loop bodies like this into a lambda, returning `true` if you've consumed it and it should get dropped, and `false` otherwise. Afterwards you can use `annotations.applyToOperation(circuitOp)` to store it back on the op.

But your approach is very clean here, so I'm not sure if `removeAnnotations` would actually help.

Reply:
This method looks great! I'll switch to it in the spirit of using existing APIs. 

The possibility of a `signalPassFailure` internal to that is a bit weird to handle. I can kludgily use a captured `bool`. However, this is begging for a return type of arbitrary type `std::pair<A, bool>` which I guess is actually `llvm::Optional<A>`. I suppose that's turning the annotation removal into something like a fold... Complete tangent.